vendor/bat-native-ledger/src/bat/ledger/internal/gemini/gemini_authorization.cc

```diff
@@ -0,0 +1,223 @@
+/* Copyright (c) 2021 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "bat/ledger/internal/gemini/gemini_authorization.h"
+
+#include <utility>
+
+#include "base/strings/string_number_conversions.h"
+#include "bat/ledger/global_constants.h"
+#include "bat/ledger/internal/common/random_util.h"
+#include "bat/ledger/internal/gemini/gemini_util.h"
+#include "bat/ledger/internal/ledger_impl.h"
+#include "bat/ledger/internal/logging/event_log_keys.h"
+#include "crypto/sha2.h"
+
+using std::placeholders::_1;
+using std::placeholders::_2;
+using std::placeholders::_3;
+using std::placeholders::_4;
+
+namespace ledger {
+namespace gemini {
+
+GeminiAuthorization::GeminiAuthorization(LedgerImpl* ledger)
+    : ledger_(ledger),
+      gemini_server_(std::make_unique<endpoint::GeminiServer>(ledger)),
+      promotion_server_(std::make_unique<endpoint::PromotionServer>(ledger)) {}
+
+GeminiAuthorization::~GeminiAuthorization() = default;
+
+void GeminiAuthorization::Authorize(
+    const base::flat_map<std::string, std::string>& args,
+    ledger::ExternalWalletAuthorizationCallback callback) {
+  const auto wallet = ledger_->wallet()->GetWallet();
+  if (!wallet) {
+    BLOG(0, "Wallet is null");
+    callback(type::Result::LEDGER_ERROR, {});
+    return;
+  }
+
+  auto gemini_wallet = GetWallet(ledger_);
+  if (!gemini_wallet) {
+    BLOG(0, "Wallet is null");
+    callback(type::Result::LEDGER_ERROR, {});
+    return;
+  }
+
+  const auto current_one_time = gemini_wallet->one_time_string;
+
+  // We need to generate new strings as soon as authorization is triggered
+  gemini_wallet->one_time_string = util::GenerateRandomHexString();
+  const bool success = ledger_->gemini()->SetWallet(gemini_wallet->Clone());
+
+  if (!success) {
+    callback(type::Result::LEDGER_ERROR, {});
+    return;
+  }
+
+  auto it = args.find("error_description");
+  if (it != args.end()) {
+    const std::string message = args.at("error_description");
+    BLOG(1, message);
+    if (message == "User does not meet minimum requirements") {
+      callback(type::Result::NOT_FOUND, {});
+      return;
+    }
+
+    callback(type::Result::LEDGER_ERROR, {});
+    return;
+  }
+
+  if (args.empty()) {
+    BLOG(0, "Arguments are empty");
+    callback(type::Result::LEDGER_ERROR, {});
+    return;
+  }
+
+  std::string code;
+  it = args.find("code");
+  if (it != args.end()) {
+    code = args.at("code");
+  }
+
+  if (code.empty()) {
+    BLOG(0, "Code is empty");
+    callback(type::Result::LEDGER_ERROR, {});
+    return;
+  }
+
+  std::string one_time_string;
+  it = args.find("state");
+  if (it != args.end()) {
+    one_time_string = args.at("state");
```

Comment:
Can or should state length be checked?

Reply:
checks here: https://github.com/brave/brave-core/pull/8566/files#diff-722cc8ead7db50e033579375725929455db1c723b37cab42541499c69c8ae7e9R98 - let me know if I'm missing something

Reply:
Sorry let me elaborate more, generally these fields are always 256bits (at least in our specs). Is that the case here? If so I generally would just check for anything not of that size to prevent api misuse.

Reply:
done - 64 char hex string - https://github.com/brave/brave-core/pull/8566/files#diff-722cc8ead7db50e033579375725929455db1c723b37cab42541499c69c8ae7e9R98